components/plurals/src/rules/runtime/ast.rs

```diff
@@ -0,0 +1,634 @@
+        where
+            E: de::Error,
+        {
+            Rule::from_str(rule_string).map_err(|err| {
+                de::Error::invalid_value(
+                    de::Unexpected::Other(&format!("{}", err)),
+                    &"a valid UTS 35 rule string",
+                )
+            })
+        }
+
+        fn visit_borrowed_bytes<E>(self, rule_bytes: &'de [u8]) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+        {
+            let rule = VarZeroVec::parse_byte_slice(rule_bytes).map_err(|err| {
+                de::Error::invalid_value(
+                    de::Unexpected::Other(&format!("{}", err)),
+                    &"a valid UTS 35 rule byte slice",
+                )
+            })?;
+            Ok(Rule(rule))
+        }
+    }
+
+    impl<'de: 'data, 'data> Deserialize<'de> for Rule<'data> {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            if deserializer.is_human_readable() {
+                deserializer.deserialize_str(DeserializeRule)
+            } else {
+                deserializer.deserialize_bytes(DeserializeRule)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::rules::reference::ast;
+    use crate::rules::reference::parse;
+    use crate::rules::runtime::test_rule;
+    use crate::PluralOperands;
+
+    #[test]
+    fn simple_rule_test() {
+        let input = "i = 1";
+        let full_ast = parse(input.as_bytes()).unwrap();
+        assert_eq!(
+            full_ast,
+            ast::Rule {
+                condition: ast::Condition(vec![ast::AndCondition(vec![ast::Relation {
+                    expression: ast::Expression {
+                        operand: ast::Operand::I,
+                        modulus: None,
+                    },
+                    operator: ast::Operator::Eq,
+                    range_list: ast::RangeList(vec![ast::RangeListItem::Value(ast::Value(1))])
+                }])]),
+                samples: None,
+            }
+        );
+
+        let rule = Rule::from(&full_ast);
+        let relation = rule
+            .0
+            .iter()
+            .next()
+            .expect("Should have a relation")
+            .as_relation();
+        assert_eq!(
+            relation,
+            Relation {
+                and_or: AndOr::And,
+                polarity: Polarity::Positive,
+                operand: Operand::I,
+                modulo: 0,
+                range_list: ZeroVec::Borrowed(&[RangeOrValue::Value(1).as_unaligned()])
+            }
+        );
+
+        let fd = fixed_decimal::decimal::FixedDecimal::from(1);
+        let operands = PluralOperands::from(&fd);
+        assert!(test_rule(&rule, &operands),);
+    }
+
+    #[test]
+    fn complex_rule_test() {
+        let input = "n % 10 = 3..4,9 and n % 100 != 10..19,70..79,90..99 or n = 0";
```

Comment:
Request: add a roundtrip test (into and out of VarULE) for a complex rule like this one